### ISSUE TITLE
Improve `/-/healthy` endpoint to check more than the webHandler

### ISSUE THREAD

**Issue by Thib17** [prometheus/prometheus#3650]
Prometheus already provides an `/-/healthy` endpoint (https://github.com/prometheus/prometheus/blob/master/web/web.go#L262). But this endpoint keep responding 200 for as long as the webHandler is up. In some situation the webHandler is up, but the Prometheus doesn't work as expected (e.g. it can't write on disk, some important mutex are deadlocked, ...).

Therefore the `/-/healthy` should be improved to better check Prometheus health in order to detect when it needs to get killed. In my opinion testing that we can at least read samples that are fresh enough could be a good start.

I would be happy to help on that, but I have no idea how should I start.

**Comment by discordianfish** [comment]
We have both /-/healthy which does what you describe and /-/ready which is doing what I think you want, right? See this for context: https://github.com/prometheus/prometheus/pull/2997

**Comment by Thib17** [comment]
`/-/ready` seems to be only used at start-up. AFAIK, once the webHandler is set ready, it never change. 
So if the Prometheus properly starts and then stop working as expected after a while, both current `/-/healthy` and `/-/ready` will keep respond 200. In such situation I believe that `/-/ready` is right to respond 200, but `/-/healthy` should respond 5xx.

**Comment by discordianfish** [comment]
If I read this correctly, it should always check if it's ready before returning 200 on the ready endpoint. The handler is wrapped by https://github.com/prometheus/prometheus/pull/2997/files#diff-73cbd2009b0e4f156e8ca0f47e95b016R273

**Comment by Thib17** [comment]
True, but as I read it `h.ready` is never set to 0. So once it's set to 1, isReady() always return true, and `/-/ready` always responds 200. And this looks normal to me, as it checks that start-up is done.

What doesn't look normal to me, is that `/-/healthy` always return 200, no matter if the Prometheus works as expected or not.

**Comment by discordianfish** [comment]
/-/ready is suppose to respond with 200 no matter what. It can be used as liveness probe in kubernetes: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

Now regarding /-/healthy, it looks like you're right. Nothing sets that to non-ready. So I suppose it's a reasonable request to have this return non-2xx if something breaks after startup. @beorn7 / @grobie Thoughts?

**Comment by grobie** [comment]
Sounds good. What and how do you want to check?

**Comment by Thib17** [comment]
I would check that reads and writes to the local storage are working.
To do that I first thought to query `up` metric and check that timestamps are fresh enough. But it doesn't work because a Prometheus without any target (without `up` metric) would never be healthy.

So I'm looking for new ideas. Do you have any suggestions ?

**Comment by brian-brazil** [comment]
Those could get expensive on large setups.

**Comment by iksaif** [comment]
@brian-brazil any idea of calls that would be good enough but still cheap ?

I think we should use the approach used in the push_gateway: https://github.com/prometheus/pushgateway/blob/c62e6bb458ff2192bc5df82a1a7f9d8ac826fac7/handler/misc.go#L23 and add `Healthy() error` to important subsystems

* ScrapeManager
* TargetManager
* RuleManager
* QueryEngine
* Notifier

The most basic implementation for these would just lock/unlock the mutex if there is one and check that there isn't obvious errors.

**Comment by roidelapluie** [comment]
I would love to have something relevant in /-/healthy.

Just do a commit to TSDB if there are len(scrape_configs/rules) > 0 ?

**Comment by beorn7** [comment]
Sounds concrete enough. I re-open right now and assign it to you.

**Comment by krajorama** [comment]
Hello from the bug scrub!

@roidelapluie implemented https://prometheus.io/docs/prometheus/latest/querying/api/#notifications API, maybe that could be hooked into here somehow? Would you be interested in adding something like https://github.com/prometheus/prometheus/issues/3650#issuecomment-785153748 ? 

krajo: I think with a concrete definition of what we want to see here, we could have community contributors do this. 

@machine424 mentioned he'll take another look.
